### PR TITLE
Fix validation of  references to CA certificate in TLS config

### LINKED ISSUE #1114: [BUG] operator crashes if tls config is invalid
If tls config is invalid,
the operator just crashes instead of giving a proper validation error.

```markdown
2022-09-09T16:22:38.660Z	DEBUG	controllers/replica_set_controller.go:140	Validating MongoDB.Spec	{"ReplicaSet": "..."}
2022-09-09T16:22:38.661Z	DEBUG	controllers/replica_set_controller.go:149	Ensuring the service exists	{"ReplicaSet": "..."}
2022-09-09T16:22:38.864Z	DEBUG	agent/replica_set_port_manager.go:122	No port change required	{"ReplicaSet": "..."}
2022-09-09T16:22:38.874Z	INFO	controllers/replica_set_controller.go:464	Create/Update operation succeeded	{"ReplicaSet": "...", "operation": "updated"}
2022-09-09T16:22:38.874Z	INFO	controllers/mongodb_tls.go:39	Ensuring TLS is correctly configured	{"ReplicaSet": "..."}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13d202e]

goroutine 354 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:118 +0x1f4
panic({0x15247e0, 0x24553c0})
	/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/mongodb/mongodb-kubernetes-operator/api/v1.MongoDBCommunity.TLSConfigMapNamespacedName(...)
	/workspace/api/v1/mongodbcommunity_types.go:797
github.com/mongodb/mongodb-kubernetes-operator/controllers.getCaCrt({_, _}, {_, _}, {{{0x14080c8, 0x10}, {0xc0004b61e0, 0x1f}}, {{0xc0003f9d20, 0x10}, ...}, ...})
	/workspace/controllers/mongodb_tls.go:166 +0x12e
github.com/mongodb/mongodb-kubernetes-operator/controllers.(*ReplicaSetReconciler).validateTLSConfig(_, {{{0x14080c8, 0x10}, {0xc0004b61e0, 0x1f}}, {{0xc0003f9d20, 0x10}, {0x0, 0x0}, {0xc0003f9d18, ...}, ...}, ...})
	/workspace/controllers/mongodb_tls.go:42 +0x125
github.com/mongodb/mongodb-kubernetes-operator/controllers.ReplicaSetReconciler.Reconcile({{0x19c5b60, 0xc0005248a0}, 0xc000158540, 0xc0003fe158, 0xc0003fe370, 0xc0003fe378}, {0x19bd600?, 0xc000130d20?}, {{{0xc0003f9d18, 0x8}, ...}})
	/workspace/controllers/replica_set_controller.go:158 +0x7d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x19bd558?, {0x19bd600?, 0xc000130d20?}, {{{0xc0003f9d18?, 0x1627a00?}, {0xc0003f9d20?, 0x4041f4?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0004a28c0, {0x19bd558, 0xc000526f00}, {0x1574ca0?, 0xc000348060?})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320 +0x33c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0004a28c0, {0x19bd558, 0xc000526f00})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:230 +0x325
```

```yaml
apiVersion: mongodbcommunity.mongodb.com/v1
kind: MongoDBCommunity
metadata:
  name: ...
spec:
  members: 1
  type: ReplicaSet
  version: "5.0.6"
  security:
    tls:
      enabled: true
      certificateKeySecretRef:
        name: mongodb-cert-secret
      #missing caConfigMapRef
```

### PR DESCRIPTION
**Pull request by adamliesko**
Closes #1114 and #1054 on GH.

### All Submissions:

* [x] Have you opened an Issue before filing this PR?

The issue has existed beforehand.

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?

Yes.

* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

#1115 Exists, however, is incomplete and is mixing the TLS validations into an (at least in my opinion) undesired place of different validations. But maybe you/we would like to actually broaden the scope of validations, happy to hear your preferred choice.

* [x] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such).

### CODE REVIEW

File: controllers/replicaset_controller_test.go [hunk 1]
```diff
@@ -87,6 +87,15 @@ func newScramReplicaSet(users ...mdbv1.MongoDBUser) mdbv1.MongoDBCommunity {
 }
 
 func newTestReplicaSetWithTLS() mdbv1.MongoDBCommunity {
+	return newTestReplicaSetWithTLSCaCertificateReferences(&mdbv1.LocalObjectReference{
```
Comment (slaskawi) [inline controllers/replicaset_controller_test.go:90]: Could you please help me understand what this change bring to the table? I'm not sure if it changes anything?

File: controllers/mongodb_tls_test.go [hunk 2]
```diff
@@ -328,7 +328,69 @@ func TestPemSupport(t *testing.T) {
...
+		assert.NoError(t, err)
+
+		r := NewReconciler(mgr)
+
+		_, err = r.validateTLSConfig(mdb)
+		assert.Nil(t, err)
+	})
+	t.Run("Failure if reference to CA cert is missing", func(t *testing.T) {
+		mdb := newTestReplicaSetWithTLSCaCertificateReferences(nil, nil)
+
```
Comment (slaskawi) [inline controllers/mongodb_tls_test.go:382]: There's quite a lot of code duplication within this test. Can I ask you to re-implement it to the parameters-driven test, like https://github.com/mongodb/mongodb-kubernetes-operator/blob/master/test/e2e/replica_set_arbiter/replica_set_arbiter_test.go#L28 ?

**Review by slaskawi — CHANGES_REQUESTED** [review]

File: controllers/replicaset_controller_test.go (hunk 1 above)
Comment (adamliesko) [reply controllers/replicaset_controller_test.go:90]: It allows us to create test fixtures, where the caller passes custom arguments.

File: controllers/mongodb_tls_test.go (hunk 2 above)
Comment (adamliesko) [reply controllers/mongodb_tls_test.go:382]: Yes sure. I initially started it as a table drive test, but skimming through the codebase it didn't seem to me like a prevalent pattern!

**Comment by slaskawi** [comment]
Integrated, thanks @adamliesko

**Comment by dan-mckean** [comment]
Hey @adamliesko - I'm the PM for our MongoDB Kubernetes operators. 

Wondered if you or any of your colleagues would be up for a chat about how you're getting on running MongoDB in Kubernetes? 

dan.mckean@mongodb.com

**Pull request merged**